src/us/kbase/workspace/test/controllers/workspace/WorkspaceController.java

```diff
@@ -0,0 +1,115 @@
+package us.kbase.workspace.test.controllers.workspace;
+
+import static us.kbase.common.test.controllers.ControllerCommon.findFreePort;
+import static us.kbase.common.test.controllers.ControllerCommon.makeTempDirs;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.commons.io.FileUtils;
+import org.ini4j.Ini;
+import org.ini4j.Profile.Section;
+
+import us.kbase.workspace.WorkspaceServer;
+
+
+/** Q&D Utility to run a Workspace server for the purposes of testing from
+ * Java. Expected to be packaged in a test jar with all dependencies.
+ * 
+ * Initializes a GridFS backend and does not support handles, byte streams, or listeners.
+ * @author gaprice@lbl.gov
+ *
+ */
+public class WorkspaceController {
+
+	private final static String DATA_DIR = "temp_data";
+	private static final String WS_CLASS = WorkspaceServer.class.getName();
+	private static final String JAR_PATH = WorkspaceServer.class.getProtectionDomain()
+			.getCodeSource().getLocation().getPath();
+
+	private final static List<String> tempDirectories = new LinkedList<String>();
+	static {
+		tempDirectories.add(DATA_DIR);
+	}
+
+	private final Path tempDir;
+
+	private final Process workspace;
+	private final int port;
+
+	public WorkspaceController(
+			final String mongoHost,
+			final String mongoDatabase,
+			final String mongoTypeDatabase,
+			final String adminUser,
+			final URL authServiceRootURL, // expected to include /testmode endpoint
+			final Path rootTempDir)
+			throws Exception {
+		tempDir = makeTempDirs(rootTempDir, "WorkspaceController-", tempDirectories);
+		port = findFreePort();
+		System.out.println("Using classpath " + JAR_PATH);
+		
+		final Path deployCfg = createDeployCfg(
+				mongoHost, mongoDatabase, mongoTypeDatabase, authServiceRootURL, adminUser);
+		final List<String> command = new LinkedList<String>();
+		command.addAll(Arrays.asList("java", "-classpath", JAR_PATH, WS_CLASS, "" + port));
+		final ProcessBuilder servpb = new ProcessBuilder(command)
+				.redirectErrorStream(true)
+				.redirectOutput(tempDir.resolve("workspace.log").toFile());
+
+		final Map<String, String> env = servpb.environment();
+		env.put("KB_DEPLOYMENT_CONFIG", deployCfg.toString());
+
+		workspace = servpb.start();
+		// TODO TEST add periodic check w/ exponential backoff
+		Thread.sleep(5000);
```

Comment:
Is this just to make sure that connections get made properly? Are there asynchronous things (like  `servpb.start()`) that go off in this time that are more of a pain to monitor?

Reply:
The workspace takes a second or two to start up after `ProcessBuilder.start()` is called to check all the connections to other services. This just waits until it should be ready to accept connections.

Reply:
I could do a microoptimization by calling workspace.ver() in a loop waiting for it to come up as per the TODO but it doesn't really seem worth if for now